Use _attr_ and add radar_type service for Environment Canada

gwww opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->



## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add select entity to be able to select radar type. Also, small tidy up using `_attr_extra_state_attributes`.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/20060

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

gwww committed:
Use write state to save state of select entity.

gwww commented:
Any chance someone could take a look at this one? Thx!

gwww commented:
I read your review. I have no idea what you are asking to be changed in this PR. This PR adds a service and moves a extra_state_attribute to the _attr_ version. I don't understand what should be changed.

frenck commented:
So, lets say, this service is removed and instead, we add a select entity that allows the user just to select the radar type, straight from the interface?

That way:
A) We don't need to add this service
B) We don't need to keep the attribute
C) Very user-friendly and instantly re-usable on any dashboard or automation (including device automations).

Additionally, if you flash that select entity as `config` it will show up and a service configuration in the UI as well.

gwww commented:
I added the select entity. Great suggestion! If you could take a look at it now that would be wonderful. I will remove the service once I've got the thumbs up on the new entity.

gwww committed:
Add select.py to .coveragerc

gwww commented:
@frenck would you have a quick look at the new select entity. If I'm on the right path I will remove the service. Thank you!

gwww committed:
Remove service; switched to select entity.

frenck commented:
>  If I'm on the right path I will remove the service.

I think you are! 👍

gwww commented:
I've finishing pulling out the service and linting. Ready for review again. Thx!

gwww commented:
Any chance someone could look at this?

frenck commented:
Please be patient.

frenck reviewed (commented):
I've just tested this locally, and the selected radar type isn't persistent?

gwww commented:
The selected type was not persistent by design. I idea was originally to have a service and that would be driven by an automation. Switching to a select entity was following the same design.

If the design were to change, and I'm OK if it does, I don't know how to persist a select entity's value. Is that the right choice? Could you point me to something else that does this?

frenck commented:
> If the design were to change, and I'm OK if it does, I don't know how to persist a select entity's value. Is that the right choice?

Well, basically it means it's a virtual entity in that case, which is not something we allow in general.

An alternative approach would be using an option flow for the config entry, so it can be configured on the integration entry-level.

gwww commented:
An option makes sense. I'll pursue that direction. Is it allowed to update an option value from a select entity?

frenck commented:
>  Is it allowed to update an option value from a select entity?

No, as that would not be an entity provided by the device or service (thus virtual and not allowed).

gwww commented:
What are the ways, if any, that an option can be updated that is not through the UI? Or, do I switch back to the original design and use a service?

The intent of the service is that people could automate switching their radar at different times of the year. Winter doesn't some at the same time or (usually) at all for some parts of Canada. The "persistence" per se was the saved automation.

frenck commented:
> Or, do I switch back to the original design and use a service?

Would have the same problem as the select entity, it is not persistent.

If I may propose a completely different angle: Why not add a camera entity for each different map?

gwww commented:
Let me chat with @michaeldavie and we will explore options. Thanks again for the help! I'm going to mark this as draft while we figure out the next steps.

gwww commented:
Hi @frenck, would using `async_get_last_state` be an acceptable solution? The bigger picture is that the camera entity would have a state attribute recording the camera type. That attribute would be populated from data in the environment canada lib and there would be a way to set it, probably from a select entity. The state would be restored in `async_added_to_hass`.

homeassistant commented:
Hello @gwww,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`2d9610b7b09faa8d497377f2e6bd4de7d94754f8`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;58657&#x2F;commits&#x2F;2d9610b7b09faa8d497377f2e6bd4de7d94754f8) - This commit has something that looks like an email address (gwwaters@gmail.com). Maybe try linking that to GitHub?.
- [`c281c44b0889ba808ceac5de205720868557e48d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;58657&#x2F;commits&#x2F;c281c44b0889ba808ceac5de205720868557e48d) - This commit has something that looks like an email address (gwwaters@gmail.com). Maybe try linking that to GitHub?.
- [`10bb55844b06b8ee95e2658d608a71ec0a53c06e`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;58657&#x2F;commits&#x2F;10bb55844b06b8ee95e2658d608a71ec0a53c06e) - This commit has something that looks like an email address (gwwaters@gmail.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

thecode committed:
Fix Shelly EM/3EM invalid energy value after reboot (#68052)

gjohansson-ST committed:
Add sensors to Sensibo for motion sensor (#67748)

bdraco commented:
> Hi @frenck, would using `async_get_last_state` be an acceptable solution? 

At least from my perspective its still a configuration option / virtual entity.

> The bigger picture is that the camera entity would have a state attribute recording the camera type. That attribute would be populated from data in the environment canada lib and there would be a way to set it, probably from a select entity. The state would be restored in `async_added_to_hass`.

If you have multiple Cameras you can do something like we do in the unifi protect integration.

We create a camera for each channel 
https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/camera.py#L64

but only enable the default one
https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/camera.py#L120

This avoids the need for an extra attribute since you can tell which one is which by the name.  If the user wants to change the resolution/channel they can enable/disable the ones they want/don't want.

gwww commented:
I want to check that I'm understanding.

What I'm hearing is have a `radar_type` extra state attribute on the camera and restore state on the camera entity. I started with that solution and quickly realized that there is no easy way to _immediately_ know that the `radar_type` in the underlying lib has changed. On the next refresh I would be able to tell by comparing the existing `radar_type` stored in `extra_state_attributes` with the existing `radar_type` in the underlying lib and at that point do a save state. I didn't like that solution as the save state happens at a future point on entity refresh.

So, the solution that is in the PR now feels a whole lot better. Everything to do with the state of the `select_entity` is encapsulated there. The save state, restore state, and the functionality to change the `radar_type`. The `select_entity` is completely decoupled from the `camera`.

What else? An config option on the camera could work. Other - I don't understand the reasoning around virtual entity and I don't understand why the `select_entity` is considered a virtual entity.

bdraco committed:
Remove unneeded permissions check from subscribe entities (#68044)

bdraco commented:
> I don't understand the reasoning around virtual entity and I don't understand why the `select_entity` is considered a virtual entity.

The select entity is a configuration option.  The state is stored in the integration itself and not the underlying api/device.

bdraco commented:
> What I'm hearing is have a `radar_type` extra state attribute on the camera and restore state on the camera entity. I started with that solution and quickly realized that there is no easy way to _immediately_ know that the `radar_type` in the underlying lib has changed. On the next refresh I would be able to tell by comparing the existing `radar_type` stored in `extra_state_attributes` with the existing `radar_type` in the underlying lib and at that point do a save state. I didn't like that solution as the save state happens at a future point on entity refresh.

I think we aren't on the same page here.  I'm suggesting creating a camera for each radar type and making the non-default ones disabled by default.  The user can then enable/disable the ones that they want

balloob committed:
Fix switch light adding itself to devices (#68060)

gwww commented:
> I think we aren't on the same page here. I'm suggesting creating a camera for each radar type and making the non-default ones disabled by default. The user can then enable/disable the ones that they want

We have avoided creating multiple entries as we do not want to enable users of the integration to potentially have two enabled radar entries. We did a lot of work to reduce load on Environment Canada servers and would like to keep the load minimal. 

Since we can't store the `select_entity`'s state I solution I'll propose is this: mod the lib to support a callback when the radar type is changed. In the camera entity, register the callback, add `radar_type` `extra_state_attribute`, and when the callback is called it will save state. State would also be restored in the camera entity.

What's the feeling around that proposal?

bdraco commented:
With this proposal now is the radar type changed, and where does it get stored ?

mfugate1 committed:
Migrate SleepIQ unique IDs that are using sleeper name instead of sleeper ID (#68062)

balloob committed:
Bump frontend to 20220312.0 (#68061)

gwww commented:
> With this proposal how is the radar type changed, and where does it get stored ?

Radar type is changed in the underlying lib using a select entity, as it is in the PR at them moment. The lib would call the registered callback in the camera entity when the value changes, which would save the state.

bdraco commented:
> camera entity when the value changes, which would save the state.

AFAICT, that still fit the definition of a virtual entity as the state is stored in the integration itself and not the underlying api/device.

gwww commented:
The radar type attribute value is stored in the underlying lib, no? I'm admit I'm totally confused. The model for what I'm trying to accomplish is from the code in alarm control panel in Elk -- it feels like what I'm doing is the same as that integration. If you feel that one is OK, then what I'll do is update this PR so we talking the real language of code! Before I do that, are there any concerns using select entity to set the radar type in the lib?

gwww commented:
Is there an ADR on virtual entities, or something I can read.

bdraco commented:
> Is there an ADR on virtual entities, or something I can read.

I don't think it comes up often enough that we have it documented.  

Here is a good test:

If you delete the integration and set it back up again and the entity loses state, and does eventually return to the expected state without user interaction, then it is very likely considered a virtual entity since the state of it is stored by Home Assistant.

SteveEasley committed:
Add remote platform to Kaleidescape integration (#67959)

Commit:
[ci skip] Translation update

gwww commented:
That def'n helps! By that def'n what I'm doing is considered virtual then. I don't understand why virtual entities should be avoided. The name "virtual entity" doesn't mean much to me to connect to the idea either. Moving on...

Back to basics... What I'm trying to do is make is easy for the user to switch the type of radar that is displayed. The two real choices are snow and rain... which depending on where you live in Canada get switched at different times of the year. On restart of HA I'd like for whatever was last selected to be the radar type displayed. There is a third setting called "Auto" that switches on two hard coded dates embedded in the lib.

One more desire, is that the switch between snow/rain can be automated. Likely something such as two user dates would drive the switch. The automation would be triggered on HA start, at least that is one way. 

I think we've established:

- Using select entity and saving it's state is not a choice (virtual entity)
- Using camera entity and saving it's state is not a choice (virtual entity by the definition, but I don't understand why its bad
- Creating both rain and snow radar is a choice we don't want to entertain because of the hit on the Environment Canada servers
- A `config_flow` option? Is that a choice? It cannot be automated that I'm aware of.
- A service against the camera to set the radar type. This is where this PR started. This would support automation.

I can't think of other choices.

Oh, yah, thanks for all the interaction!!!

bdraco commented:
> I don't understand why virtual entities should be avoided

At least one reason is that every time you login to Home Assistant the contents of the state machine gets downloaded into your browser/device. We want to avoid creating entities that are solely internal configuration state as if we started doing this for ever integration the state machine would quickly run into scale problems (which we already have.. see https://github.com/home-assistant/core/commit/0d8f649bd65c8c54cd3503dd75485d3ec35d6076).

bdraco commented:
> * Creating both rain and snow radar is a choice we don't want to entertain because of the hit on the Environment Canada servers

For clarity, I still thing this is a good choice if you create them as disabled by default so the user has to go in an enable them.  You could have an `Auto` camera which is enabled by default and a `Rain` and `Snow ` camera which is disabled by default.  The user could then turn on the ones they want.   I know you are concerned about the user having both, but AFAICT there isn't anything stopping them from setting up multiple copies of the integration.

bdr99 committed:
Add switch platform to the Mazda integration (#68025)

michaeldavie commented:
I'm totally confused. Why can we not simply have a single entity with an attribute that the user can control via a service? I don't see why creating three different entities that only differ by an attribute could make more sense. How is this any different than any other entity that's configurable?

bdraco commented:
Configuration data should be stored in the config entry data or options section. It should not be stored in the state machine

michaeldavie commented:
This isn’t a static configuration setting though, it’s a dynamic one. Ideally it would be automatically adjusted according to the current weather conditions without user intervention.

To put it another way, the user selecting what precipitation type they want the camera to display at a given moment is analogous to them choosing a song for a media player. Creating a different entity for every possible option doesn’t make sense to me.

bdraco commented:
As long as you aren't storing it in the state machine, that seems fine

allenporter committed:
Update google-cloud-pubsub to 2.11.0 (#68074)

cthornton committed:
Default somfy_mylink shade's _attr_is_closed to `None` (#68053)

The base `Cover` entity requires an explicit value for
`_attr_is_closed`.

Since the `SomfyShade` is an assumed state, we don't know
by default whether the shade is open or not, so we need to
explicitly return `None` for `_attr_is_closed`

frenck committed:
Add base entity to Switch as X (#68057)

frenck committed:
Update sqlalchemy to 1.4.32 (#68075)

frenck committed:
Update sentry-sdk to 1.5.7 (#68077)

frenck committed:
Update google-cloud-texttospeech to 2.11.0 (#68076)

frenck committed:
Update wled to 0.13.1 (#68084)

BraveChicken1 committed:
Update home_connect to 0.7.0 (#68089)

frenck committed:
Fix Efergy tests (#68086)

bdraco committed:
Bump pyisy to 3.0.5 (#68069)

* Bump pyisy to 3.0.4

- Fixes #66003

- Changelog: https://github.com/automicus/PyISY/compare/v3.0.1...v3.0.4

* again

gwww commented:
I see this as down to adding a service to change the radar type and being able to have automations to call the service as the end-user wants. This allows them to automate according to however they want. Radar type is not persisted.

The other choice is an option on config flow, which AFAIK cannot be automated and that is a big downside.

The final choice is do nothing, which is what I'm going to do if neither of the above two options are acceptable.

Would adding the service be accepted? If so I'll restore that commit and resubmit.

mfugate1 committed:
Clean up SleepIQ migration (#68092)

frenck committed:
Small improvements for Switch as X (#68079)

* Small improvements forr Switch as X

* Test improvements

* Remove intregration tests from config flow test

frenck committed:
Update freezegun to 1.2.0 (#68090)

CoMPaTech committed:
Update plugwise module to 0.16.8 (#68082)

Junker committed:
Update slixmpp to 1.8.0.1 (#68080)

fix for python 3.10

bdraco commented:
> I see this as down to adding a service to change the radar type and being able to have automations to call the service as the end-user wants. This allows them to automate according to however they want. Radar type is not persisted.

Seems fine to me as long as we don't store it in the state machine

frenck committed:
Remove unused types for emoji (#68091)

Commit:
[ci skip] Translation update

gjohansson-ST committed:
Add Pure devices to Sensibo (#67695)

bdraco committed:
Remove unused columns from states/events tables (#68078)

gjohansson-ST committed:
Add binary_sensor platform for Sensibo (#68088)

jeeftor committed:
Add intellifire UDP discovery at configuration start (#67002)

Co-authored-by: J. Nick Koston <nick@koston.org>

frenck committed:
Add fixture for enabling disabled entities in WLED tests (#68087)

frenck committed:
Update pytest to 7.1.0 (#68108)

frenck committed:
Update pyupgrade to v2.31.1 (#68110)

frenck committed:
Clean up Whois tests use entity_registry_enabled_by_default fixture (#68113)

frenck committed:
Only create WLED current sensors when available (#68116)

frenck committed:
Add cover platform to Switch as X (#68107)

jbouwh committed:
Fix MQTT false positive deprecation warnings (#68117)

frenck committed:
Add siren platform to Switch as X (#68118)

* Add siren platform to Switch as X

* Clarify inline test documentation

* Clarify inline test documentation

frenck committed:
Add Fan platform to Switch as X (#68122)

frenck committed:
Remove deprecated OpenZWave integration (#68054)

flacjacket committed:
Bump amcrest version to 1.9.7 (#68055)

Kane610 committed:
Add typing of deCONZ device_trigger (#67496)

mfugate1 committed:
Set isolated_build = True for tox (#67238)

Kane610 committed:
Add typing of deconz_event (#67497)

frenck committed:
Add Lock platform to Switch as X (#68123)

Co-authored-by: Erik Montnemery <erik@montnemery.com>

thecode committed:
Downgrade SSDP failed to setup listener warning to debug (#68129)

Kane610 committed:
Complete typing of some platforms of deCONZ integration (#67494)

GidoHakvoort committed:
Bump Rova to 0.3.0 (#67688)

bdraco committed:
Fix Magic Home devices with multiple network interfaces (#68029)

emontnemery committed:
Hide switch_as_x tracked entity (#67949)

* Hide switch_as_x tracked entity

* Hide wrapped switch during config flow

* Allow setting/getting entity disabled by via WS

* Adjust tests

* Improve test coverage

* Improve tests

balloob committed:
Add media source support to openhome (#67566)

frenck committed:
Update pytest-xdist to 2.5.0 (#68135)

emontnemery committed:
Restore state of trigger based template binary sensor (#67538)

balloob committed:
Add media source support to unifiprotect (#67570)

bdraco committed:
Ensure WiZ is reloaded on title change (#68147)

bdraco committed:
Ensure flux_led is reloaded when the title changes (#68146)

emontnemery committed:
Minor tweak of config entity_registry test (#68141)

bbr111 committed:
Bump volkszahler to 0.3.2 (#67958)

Co-authored-by: Franck Nijhof <git@frenck.dev>

allenporter committed:
Update google calendar integration with a config flow (#68010)

* Convert google calendar to config flow and async

* Call correct exchange method

* Fix async method and reduce unnecessary diffs

* Wording improvements

* Reduce unnecessary diffs

* Run load/update config from executor

* Update homeassistant/components/google/calendar.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Remove unnecessary updating of unexpected multiple config entries.

* Remove unnecessary unique_id checks

* Improve readability with comments about device code expiration

* Update homeassistant/components/google/calendar.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Update homeassistant/components/google/calendar.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Update homeassistant/components/google/api.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Add comment for when code is none on timeout

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

epenet committed:
Handle ConnectionClosed in SamsungTV try_connect (#68125)

* Handle ConnectionClosed in SamsungTV try_connect

* Add tests

* Add quotes around the error message

Co-authored-by: epenet <epenet@users.noreply.github.com>

cdce8p committed:
Remove unused TypeVars (#68155)

droberg committed:
Clean up code for onewire config flow (#67970)

* Change debug level of integration reload after config update to debug

* Remove extra .keys() call for dict itreation

* Remove unecessary type check

* Remove unused LOGGER reference

CoMPaTech committed:
Prevent adding plugwise products used as secondary controllers (#68098)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cdce8p committed:
Improve sonos error decorator typing (#67199)

cdce8p committed:
Don't import TypeVars from core modules (#68154)

Kane610 committed:
Fix deconz typing (#68143)

nprez83 committed:
Fix lyric climate (#67018)

* Fixed the issues related to auto mode

I was having the same issues as described in #63403, specifically, the error stating that Mode 7 is not valid, only Heat, Cool, Off when trying to do anything while the thermostat is set to Auto. This error originates with the way the Lyric API handles the modes. Basically, when one queries the changeableValues dict, you get a mode=Auto, as well as a heatCoolMode, which is set to either Heat, Cool, Off. Per the documentation, heatCoolMode contains the "heat cool mode when system switch is in Auto mode". It would make sense that when changing the thermostat settings, mode=Auto should be valid, but it's not. The way the API understands that the mode should be set to Auto when changing the thermostat settings is by setting the autoChangeoverActive variable to true, not the mode itself. This require changes in the async_set_hvac_mode, async_set_temperature, and async_set_preset_mode functions. Related to this issue, I got rid of the references to hasDualSetpointStatus, as it seems that it always remains false in the API, even when the mode is set to auto, so again, the key variable for this is autoChangeoverActive.

While I was working on this I also noticed another issue. The support flag SUPPORT_TARGET_TEMPERATURE_RANGE had not been included, which did not allow for the temperature range to be available, thus invalidating the target_temperature_low and target_temperature_high functions. I added this flag and sorted out which set point (heat vs cool) should be called for each of them so things work as expected in Lovelace. I have tested all of these functionalities and they all work great on my end, so I thought I'd share.

* Update climate.py

* Update climate.py

Fixed two additional issues: 1) When the system is turned off from Auto, the heatCoolMode variable becomes 'Off', so when you try to restart the system back to Auto, nothing happens. 2) I now prevent the async_set_temperature function from being called with a new set point when the system is Off.
All changes tested and functional.

* Update climate.py

* Update climate.py

Return SUPPORT_PRESET_MODE flag only for LCC models (i.e. they have the "thermostatSetpointStatus" variable defined). TCC models do not support this feature

* Update climate.py

After playing with the official Honeywell API, I realized it doesn't like to received commands with missing data, i.e., it always wants to get a mode, coolSetpoint, heatSetpoint, and autoChangeoverActive variables. This was causing some random issues with changing modes, especially from coming from off, so I modified the async_set_temperature, and async_set_hvac_mode fuctions to always send all pertinent variables.

* Update climate.py

* Update climate.py

* Update climate.py

* Update climate.py

* Clean code and test everything

Alright, sorry for the multiple commits, fixing this properly took a fair bit of testing. I went ahead and cleaned up the code and made the following big picture changes: 
1) The integration now supports the Auto mode appropriately, to include the temperature range. 
2) There's a bug that actually manifests when using the native app. When the system is 'Off' and you try to turn it on to 'Auto', it will turn on briefly but will go back to 'Off' after a few seconds. When checking the web api, this appears to be related to the fact that the heatCoolMode variable seems to continue to store 'Off', even if the mode accurately displays 'Auto', and the autoChangeoverActive=True. So to overcome that inherent limitation, when the system is 'Off' and the user turns it to 'Auto', I first turn it to Heat, wait 3 seconds, and then turn it to 'Auto', which seems to work well.

* Update climate.py

* Fixed errors

* Fixed comments that were resulting in error.

* Update climate.py

* Update homeassistant/components/lyric/climate.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Update homeassistant/components/lyric/climate.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Update climate.py

I removed a blank line in 268 and another one at the end of the document. I also fixed the outdents of await commands after the _LOGGER.error calls, not sure what else may be driving the flake8 and black errors. Any guidance is much appreciated @MartinHjelmare

* Update climate.py

* Update climate.py

corrected some indents that I think were the culprit of the flake8 errors

* Update climate.py

I used VS Code to fix locate the flake8 errors. I ran black on it, so I'm hoping that will fix the last lingering black error.

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

frenck committed:
Update flake8-comprehensions to 3.8.0 (#68164)

frenck committed:
Update twentemilieu to 0.6.0 (#68171)

emontnemery committed:
Prevent spawning script runs when shutting down (#68170)

cdce8p committed:
Fix met TypeVar usage (#68152)

* Fix met TypeVar usage

* Change format_condition

cdce8p committed:
Fix point by adding authlib constraint (#68176)

* Fix point by pinning authlib

* Use constraint

gwww committed:
Bump lib version and disable AQHI

gwww committed:
Merge branch 'ec-camera' of github.com:gwww/core into ec-camera

gwww committed:
Add service to set the type of radar to fetch.

gwww commented:
Sorry folks. Somehow, my repo got borked and this PR was rebased somewhere it should not have been. Closing and will repopen new one. I will reference this PR for the conversation history.

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#58657).

Once you do that we will be able to review and accept this pull request.

Thanks!
